packages/matter.js/src/cluster/server/AttributeServer.ts

```diff
@@ -38,216 +95,366 @@ export abstract class BaseAttributeServer<T> {
+            if (valueChanged) {
+                this.version++;
+                this.matterListeners.forEach(listener => listener(value, this.version));
+            }
+            this.listeners.forEach(listener => listener(value, oldValue));
+            return;
+        }
 
         this.setLocal(value);
     }
 
+    /**
+     * Set the value of the attribute locally. This method should be used locally in the code and does not include the
+     * ACL check.
+     * If a setter is defined this setter method is called to validate and store the value.
+     * Else if a validator is defined the value is validated before it is stored.
+     * Listeners are called when the value changes (internal listeners) or in any case (external listeners).
+     */
     setLocal(value: T) {
+        this.schema.validate(value);
+
+        if (this.setter !== undefined) {
+            if (this.getter === undefined) {
+                throw new Error("Attributes with Setter methods also need a Getter method to be readable.");
+            }
+
+            const oldValue = this.getter(undefined, this.endpoint);
+            const valueChanged = this.setter(value, undefined, this.endpoint);
+            if (valueChanged) {
+                this.version++;
+                this.matterListeners.forEach(listener => listener(value, this.version));
+            }
+            this.listeners.forEach(listener => listener(value, oldValue));
+            return;
+        }
+
+        if (this.validator !== undefined) {
+            this.validator(value, undefined, this.endpoint);
+        }
+
         const oldValue = this.value;
         if (!isDeepEqual(value, oldValue)) {
-            this.validator(value, this.name);
             this.version++;
             this.value = value;
             this.matterListeners.forEach(listener => listener(value, this.version));
         }
-        this.listeners.forEach(listener => listener(value, oldValue));
+        if (oldValue !== undefined) { // should always be the case, but let's make sure
+            this.listeners.forEach(listener => listener(value, oldValue));
+        }
+    }
+
+    /**
+     * When the value is handled by getter or setter methods and is changed by other processes this method can be used
+     * to notify the attribute server that the value has changed. This will increase the version number and trigger the
+     * listeners.
+     * ACL checks needs to be performed before calling this method.
+     */
+    updated(session: SecureSession<MatterDevice>) {
+        if (this.getter === undefined && this.setter === undefined) {
+            throw new Error("Updated can only be used if the attribute has a getter or setter method.");
+        }
+        const value = this.get(session, true);
+        this.version++;
+        this.matterListeners.forEach(listener => listener(value, this.version));
+        // We can not call the internal listeners here, because we do not know the old value.
```

Comment:
I also thought about that. In the end we would store then values twice in two places and also initialize it from persistence and such. I decided to not store it to prevent cases where it gets inconsistent then. Should ideally not happen. Additionally we need storage space twice ;-)

Reply:
Not working when fabric scoped so each value is fabric scoped normally, so each set value is potentially for another fabric ... no idea how to solve that - or we need to store 8and also init at begin ing) all values per fabric.
Lets see if we miss it. I added TODO

Reply:
changed